pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,166 +18,480 @@ import (
 	"context"
 	"fmt"
 	"math"
-	"strings"
 
-	"github.com/Pallinder/go-randomdata"
-	"github.com/mitchellh/hashstructure/v2"
+	"github.com/aws/karpenter/pkg/utils/apiobject"
+
+	"go.uber.org/multierr"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
-	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
-	"github.com/aws/karpenter/pkg/utils/functional"
 	"github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/sets"
 )
 
 type Topology struct {
 	kubeClient client.Client
+	topologies map[uint64]*TopologyGroup
+	// Anti-affinity works both ways (if a zone has a pod foo with anti-affinity to a pod bar, we can't schedule bar to
+	// that zone, even though bar has no anti affinity terms on it. For this to work, we need to separately track the
+	// topologies of pods with anti-affinity terms, so we can prevent scheduling the pods they have anti-affinity to
+	// in some cases.
+	inverseTopologies map[uint64]*TopologyGroup
+	constraints       *v1alpha5.Constraints
+}
+
+func NewTopology(kubeClient client.Client, constraints *v1alpha5.Constraints) *Topology {
+	return &Topology{
+		kubeClient:        kubeClient,
+		constraints:       constraints,
+		topologies:        map[uint64]*TopologyGroup{},
+		inverseTopologies: map[uint64]*TopologyGroup{},
+	}
 }
 
-// Inject injects topology rules into pods using supported NodeSelectors
-func (t *Topology) Inject(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) error {
-	// Group pods by equivalent topology spread constraints
-	topologyGroups := t.getTopologyGroups(pods)
-	// Compute spread
-	for _, topologyGroup := range topologyGroups {
-		if err := t.computeCurrentTopology(ctx, constraints, topologyGroup); err != nil {
-			return fmt.Errorf("computing topology, %w", err)
+func (t *Topology) TrackTopologies(ctx context.Context, pods ...*v1.Pod) error {
```

Comment:
Thoughts on this naming suggestion?
```suggestion
func (t *Topology) Update(ctx context.Context, pods ...*v1.Pod) error {
```